benchmarks/kmeans/objective.py

```diff
@@ -0,0 +1,84 @@
+from benchopt import safe_import_context
+from benchopt.base import BaseObjective
+
+with safe_import_context() as import_ctx:
+    import numpy as np
+
+
+class Objective(BaseObjective):
+    name = "Lloyd walltime"
+    url = "https://github.com/soda-inria/sklearn-engine-benchmarks"
+
+    requirements = ["numpy"]
+
+    min_benchopt_version = "1.4.0"
+
+    # Since our goal is to measure walltime for solvers that perform exact same
+    # computations, the solver parameters are part of the objective and must be set
+    # for all solvers, rather than being an independent benchmark space for each
+    # solver.
+    parameters = dict(
+        n_clusters=[127],
+        init=["random", "k-means++"],
+        n_init=[1],
+        max_iter=[100],
+        tol=[0],
+        verbose=[0],  # NB: for kmeans, verbosity can affect the performance
+        algorithm=["lloyd"],
+        random_state=[123],
+        sample_weight=["None", "unary", "random"],  # ???: defaut ?
```

Comment:
Do we really need unary? I would test either no weights or random weights, with a default to no weights.

Reply:
Some implementations (like sklearn-pytorch-engine) benefit from detecting that weights are unary to fall back to the "None" cases, where there's a different code path that skip entirely the weighting. This option is sort of a safety check that those implementations that benefit from it actually optimize this case.

Some other implementations don't bother and create unary weights anyway if sample_weight is None and use the same code path for all cases.

Reply:
Honestly, I don't see the use case for k-means with unary weights. I think we should get rid of those code paths entirely, they add maintenance complexity for no value-added for the users.

Reply:
It does make a difference for the pytorch engine. If weights must be taken into account, it cost one additional copy of the whole dataset per iteration (for the upadte of centroids), plus it has to be batched to avoid memory overflow.

Reply:
revert now